### PR TITLE
Order model

### PR DESCRIPTION
**Pull request by vfilimo**

### CODE REVIEW

File: src/main/java/online/store/book/controller/OrderController.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    private static final int DEFAULT_PAGE = 0;
+    private static final int DEFAULT_PAGE_SIZE = 5;
+    private static final String DEFAULT_SORT_PARAMETER = "id";
+    private final OrderService orderService;
+
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @Operation(summary = "Create new order by you shopping cart",
+            description = "After placing an order, your shopping cart will be cleared")
+    @PostMapping
+    OrderResponseDto placeOrder(@RequestBody @Valid OrderRequestDto orderRequestDto) {
```
Comment (Elena-Bruyako) [inline src/main/java/online/store/book/controller/OrderController.java]:
```suggestion
    public OrderResponseDto placeOrder(@RequestBody @Valid OrderRequestDto orderRequestDto) {
```

for all endpoints

File: src/main/java/online/store/book/repository/order/OrderRepository.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+            + "FROM Order o JOIN o.orderItems oi JOIN o.user u "
+            + "WHERE u.id = :userId AND o.id = :orderId")
+    @EntityGraph(attributePaths = {"orderItems", "orderItems.book"})
+    Page<OrderItem> findItemsById(@Param("userId") Long userId,
+                                 @Param("orderId") Long orderId, Pageable pageable);
+
+    @Query("SELECT oi FROM Order o JOIN o.user u JOIN o.orderItems oi "
+            + "WHERE u.id = :userId AND o.id = :orderId AND oi.id = :itemId")
+    @EntityGraph(attributePaths = {"orderItems", "orderItems.book"})
+    Optional<OrderItem> findItemById(
```
Comment (Elena-Bruyako) [inline src/main/java/online/store/book/repository/order/OrderRepository.java]: it should be in OrderItemRepo
Also, check this query without any EntityGraph 

```
@Query(value = """
            SELECT oi FROM OrderItem oi
            JOIN FETCH oi.order o
            WHERE oi.id = :orderItemId AND o.id = :orderId AND o.user.id = :userId""")
    Optional<OrderItem> findByIdAndOrderIdAndUserId(@Param("orderItemId") Long orderItemId,
                                                    @Param("orderId") Long orderId,
                                                    @Param("userId") Long userId);
```

File: src/main/java/online/store/book/service/order/OrderServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    private Order createOrder(User user, OrderRequestDto orderRequestDto,
+                              ShoppingCart shoppingCart) {
+        Order order = new Order();
+        order.setUser(user);
+        order.setStatus(Order.Status.PENDING);
+        order.setOrderDate(LocalDateTime.now());
+        order.setShippingAddress(orderRequestDto.shippingAddress());
+        Set<OrderItem> orderItems = shoppingCart.getCartItems().stream()
```
Comment (Elena-Bruyako) [inline src/main/java/online/store/book/service/order/OrderServiceImpl.java]: what if shoppingCart.getCartItems() is empty?

File: src/main/java/online/store/book/service/order/OrderServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    @Override
+    public OrderResponseDto saveOrder(User user, OrderRequestDto orderRequestDto) {
+        ShoppingCart shoppingCart = getShoppingCartByUser(user);
+        Order order = createOrder(user, orderRequestDto, shoppingCart);
+        shoppingCart.getCartItems().clear();
+        shoppingCartRepository.save(shoppingCart);
+        return orderMapper.toDto(orderRepository.save(order));
+    }
+
+    private ShoppingCart getShoppingCartByUser(User user) {
```
Comment (Elena-Bruyako) [inline src/main/java/online/store/book/service/order/OrderServiceImpl.java:78]: private methods should be placed after public methods

File: src/main/java/online/store/book/repository/order/OrderRepository.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.jpa.repository.EntityGraph;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.data.repository.query.Param;
+
+public interface OrderRepository extends JpaRepository<Order, Long> {
+    @Query("SELECT o FROM Order o JOIN o.user u WHERE u.id = :userId")
+    @EntityGraph(attributePaths = {"user", "orderItems", "orderItems.book"})
+    Page<Order> findByUserId(@Param("userId") Long userId, Pageable pageable);
```
Comment (Elena-Bruyako) [inline src/main/java/online/store/book/repository/order/OrderRepository.java]:
```suggestion
    @EntityGraph(attributePaths = "orderItems")
    List<Order> findAllByUserId(Long userId, Pageable pageable);
```

File: src/main/java/online/store/book/service/order/OrderServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+    @Override
+    public List<OrderItemResponseDto> getOrderById(User user, Long orderId, Pageable pageable) {
+        return orderItemMapper.toDto(
+                orderRepository.findItemsById(user.getId(), orderId, pageable));
+    }
+
+    @Override
+    public OrderItemResponseDto getOrderItemById(User user, Long orderId, Long itemId) {
+        OrderItem orderItem = orderRepository.findItemById(user.getId(), orderId, itemId)
```
Comment (Elena-Bruyako) [inline src/main/java/online/store/book/service/order/OrderServiceImpl.java]: Simpler 

`OrderItem item = orderItemRepository.findByIdAndOrderIdAndUserId(orderItemId, orderId, userId)`

File: src/main/java/online/store/book/service/order/OrderServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Can't find order with id: " + orderId));
+        order.setStatus(orderUpdateRequestDto.status());
+        return orderMapper.toDto(orderRepository.save(order));
+    }
+
+    @Override
+    public List<OrderItemResponseDto> getOrderById(User user, Long orderId, Pageable pageable) {
+        return orderItemMapper.toDto(
+                orderRepository.findItemsById(user.getId(), orderId, pageable));
```
Comment (Elena-Bruyako) [inline src/main/java/online/store/book/service/order/OrderServiceImpl.java]:
```suggestion
                Order order = orderRepository.findByIdAndUserId(orderId, userId)...
                return orderItemMapper.toOrderItemDtoList(order.getOrderItems());
```

where 

```
@EntityGraph(attributePaths = "orderItems")
Optional<Order> findByIdAndUserId(Long orderId, Long userId);
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/online/store/book/exceptions/EmptyCartException.java [hunk 8]
```diff
@@ -0,0 +1,11 @@
+package online.store.book.exceptions;
+
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(HttpStatus.BAD_REQUEST)
+public class EmptyCartException extends RuntimeException {
```
Comment (Elena-Bruyako) [inline src/main/java/online/store/book/exceptions/EmptyCartException.java]:
```suggestion
public class OrderProcessingException extends RuntimeException {
```

File: src/main/java/online/store/book/exceptions/EmptyCartException.java [hunk 9]
```diff
@@ -0,0 +1,11 @@
+package online.store.book.exceptions;
+
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(HttpStatus.BAD_REQUEST)
```
Comment (Elena-Bruyako) [inline src/main/java/online/store/book/exceptions/EmptyCartException.java]:
```suggestion
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Review by okuzan — APPROVED** [review]

**Pull request merged**